File: problems/3.array/ts/find-trapped-area.ts
export function findTrappedArea(heights: number[]) {
    let total = 0;
    for (
        let currentPointer = 0;
        currentPointer < heights.length;
        ++currentPointer
    ) {
        let maxRight = 0,
            maxLeft = 0,
            rightPointer = currentPointer,
            leftPointer = currentPointer;

        while (rightPointer < heights.length) {
            maxRight = Math.max.apply(null, [maxRight, heights[rightPointer]]);
            ++rightPointer;
        }

        while (leftPointer >= 0) {
            maxLeft = Math.max.apply(null, [maxLeft, heights[leftPointer]]);
            --leftPointer;
        }

        const currentWater =
            Math.min.apply(null, [maxLeft, maxRight]) - heights[currentPointer];

        if (currentWater >= 0) {
            total += currentWater;
        }
    }
    return total;
}

export function findTrappedArea2(heights: number[]) {
    let total = 0;
    let L = 0,
        R = heights.length - 1,
        maximumRight = heights[R],
        maximumLeft = heights[L];

    for (; L < R; ) {
        if (heights[L] >= heights[R]) {
            if (maximumRight <= heights[R]) {
                maximumRight = heights[R];
            } else {
                total += maximumRight - heights[R];
            }
            --R;
        } else {
            if (maximumLeft <= heights[L]) {
                maximumLeft = heights[L];
            } else {
                total += maximumLeft - heights[L];
            }
            ++L;
        }
    }

    return total;
}

File: problems/3.array/go/find_trapped_area.go
package main

import "math"

func findTrappedArea(arr []int) int {
	p1 := 0
	p2 := 1
	current := p1
	total := 0

	for p1 < p2 {
		if current == p2 {
			p1 += 1
			current = p1
			if p2 < len(arr) {
				p2 += 1
			}
		}

		if arr[p2] == 0 {
			if p2 < len(arr) {
				p2 += 1
				continue
			}
		}

		if p1 == p2 {
			break
		}

		yMax := math.Min(float64(arr[p1]), float64(arr[p2]))
		xMax := p2 - p1 - 1
		area := int(yMax) * xMax

		if area-arr[current] > 0 { // 0
			total += area - arr[current]
		}

		current += 1
	}

	return total
}
